### DISCUSSION TITLE
Time Warp Mechanic

### DISCUSSION THREAD

**Discussion by The-HandyDandyman** [Techokami/SonicWorldsNext#42] [Q&A] [answered]
Hello there! I'm working on a personal project and I wanted to ask out of sheer curiosity to see if this was even possible:
How would one go about implementing a time warp mechanic similar to the one seen in Sonic CD?
I'm mainly focusing on figuring out how after the warp to keep momentum/player state and position since I plan to make the warp instantaneous rather than based on carrying speed for long periods of time. Any help on figuring out how this would work would be greatly appreciated!

**Comment by Renhoex** [comment, marked as answer]
There's several ways to do it, I think the easiest way would to probably have the different time periods be in the scene, have 4 nodes for present, past, future and global (global would be for the player so they wouldn't be affected by time)

in the level initialization script, save the other time periods to a variable and use remove_child and add_child to switch between the periods.
so for example initializing could be
`
var getPresent = get_node("Present")
var getPast = get_node("Past")
var getFuture = get_node("Future")
func _ready():
    remove_child(getPast)
    remove_child(getFuture)
`
and you could call a function to change the time preiods something like
`
remove_child([current time period])
add_child(getFuture)
`

**Reply by The-HandyDandyman** [reply]
Thank you so much for the answer! Will make sure to test this out, I'm eager to get this up and working.

**Reply by fishingrosd113** [reply]
Hello, just popping in to ask if any of you could expand on that? What I mean is if you found a way to turn it into the signposts with the transition as well since I'm also trying to make a project that is more like Sonic CD itself. I haven't found any other sources that explain how this could be done until now, especially in Godot so I'm just really curious how it would be done there!!! Sorry for bothering I'm still fairly new to Godot and Github in general

**Reply by DimensionWarped** [reply]
> Hello, just popping in to ask if any of you could expand on that? What I mean is if you found a way to turn it into the signposts with the transition as well since I'm also trying to make a project that is more like Sonic CD itself. I haven't found any other sources that explain how this could be done until now, especially in Godot so I'm just really curious how it would be done there!!! Sorry for bothering I'm still fairly new to Godot and Github in general

The sign post and the transition are very simple implementation details that should be of your own design. There isn't any one size fits all solution for how to approach that. You can implement all of that stuff later -- for now I'd recommend starting with a simple debug key that immediately activates your time transition without regard to things like the whole rigamarole Sonic CD did with the past/future posts and all that jazz around having to stay above such and such ground speed for x amount of time and all that.

None of us have done this ourselves, so the advice Ren gave is more of a suggestion to go start tinkering from than anything.
